connection.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ocagent
+
+import (
+	"math/rand"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	sDisconnected int32 = 5 + iota
+	sConnected
+)
+
+func (ae *Exporter) setStateDisconnected() {
+	atomic.StoreInt32(&ae.connectionState, sDisconnected)
+	ae.disconnectedCh <- true
+}
+
+func (ae *Exporter) setStateConnected() {
+	atomic.StoreInt32(&ae.connectionState, sConnected)
+}
+
+func (ae *Exporter) connected() bool {
+	return atomic.LoadInt32(&ae.connectionState) == sConnected
+}
+
+const defaultConnReattemptPeriod = 5000 * time.Millisecond
```

Comment:
5 seconds seems too frequent to me, consider 5 minutes by default instead?

Reply:
5 minutes seems too long though, because even the max grpc exponential backoff delay by default is 2 minutes(120 seconds) https://godoc.org/google.golang.org/grpc#DefaultBackoffConfig but that's after trying for 1 second, about 7 times which basically becomes trying for at least 247 seconds(4.11 minutes) altogether. Perhaps a value in between?

Reply:
One minute may be a good compromise?

Reply:
Perfect, I was thinking the same too.

Reply:
On second thought, I've looked at the application sitting still for 1 minute, that seems like waaay too long even still. Am looking at perhaps 10 seconds: gRPC-Go by default does a dial for atleast every one of these times:
[1s, 2s, 4s, 8s, 16s, 32s, 64s, 120s] then gives up
putting it at 1 minute would be at the end of that time wait and can result in the loss of many spans.